src/main/java/at/hannibal2/skyhanni/features/inventory/experimentationtable/ExperimentsDryStreakDisplay.kt

```diff
@@ -0,0 +1,129 @@
+
+    private var display = emptyList<String>()
+
+    private var didJustFind = false
+
+    private val patternGroup = RepoPattern.group("enchanting.experiments.drystreak")
+    val experimentInventoriesPattern by patternGroup.pattern(
+        "inventories",
+        "(?:Superpairs|Chronomatron|Ultrasequencer) (?:\\(.+\\)|➜ Stakes|Rewards)|Experimentation Table",
+    )
+    private val enchantingExpChatPattern by patternGroup.pattern(
+        "exp",
+        "^ \\+(?<amount>\\d+|\\d+,\\d+)k? Enchanting Exp$",
+    )
+
+    @SubscribeEvent
+    fun onChestGuiOverlayRendered(event: GuiRenderEvent.ChestGuiOverlayRenderEvent) {
+        if (!isEnabled()) return
+        if (!experimentInventoriesPattern.matches(InventoryUtils.openInventoryName())) return
+
+        display = drawDisplay()
+        config.dryStreakDisplayPosition.renderStrings(
+            display,
+            posLabel = "Experiment Dry Streak Display",
+        )
+    }
+
+    @SubscribeEvent
+    fun onInventoryOpen(event: InventoryOpenEvent) {
+        if (event.inventoryName == "Experimentation Table" && didJustFind) didJustFind = false
+    }
+
+    @SubscribeEvent
+    fun onInventoryUpdated(event: InventoryUpdatedEvent) {
+        if (!isEnabled() || didJustFind) return
+
+        if (InventoryUtils.getCurrentExperiment() != null) {
+            for (lore in event.inventoryItems.map { it.value.getLore() }) {
+                val firstLine = lore.firstOrNull() ?: continue
+                if (!ultraRarePattern.matches(firstLine)) continue
+                val bookNameLine = lore.getOrNull(2) ?: continue
+                bookPattern.matchMatcher(bookNameLine) {
+                    val storage = storage ?: return
+                    ChatUtils.chat(
+                        "§a§lDRY-STREAK ENDED! §eYou have (finally) " +
+                            "found a §5ULTRA-RARE §eafter §3${storage.xpSince.shortFormat()} Enchanting Exp " +
+                            "§e and §2${storage.attemptsSince} attempts§e!",
+                    )
+                    storage.attemptsSince = 0
+                    storage.xpSince = 0
+                    didJustFind = true
+                }
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onInventoryClose(event: InventoryCloseEvent) {
+        if (didJustFind || InventoryUtils.getCurrentExperiment() == null) return
+
+        val storage = storage ?: return
+        storage.attemptsSince += 1
+    }
+
+    @SubscribeEvent
+    fun onChat(event: LorenzChatEvent) {
+        if (!isEnabled() || didJustFind) return
+
+        enchantingExpChatPattern.matchMatcher(event.message.removeColor()) {
+            val storage = storage ?: return
+            storage.xpSince += group("amount").substringBefore(",").toInt() * 1000
+        }
+    }
+
+    private fun drawDisplay() = buildList {
+        if (!config.attemptsSince && !config.xpSince) return@buildList
```

Comment:
this feels like the wrong spot to check. can we move the "is feature active" checks into onChestGuiOverlayRendered and fully return, not even calling an empty list instead?